acmwebsite/controllers/wiki.py

```diff
@@ -0,0 +1,99 @@
+"""Wiki controller module"""
+from tg import expose
+
+from acmwebsite.lib.base import BaseController
+
+from pygit2 import Repository,init_repository
+from pygit2 import Tree
+from pygit2 import Signature
+import pygit2 as pg
+
+from docutils.core import publish_parts
+
+import os
+import tg
+
+__all__ = ['WikiController']
+
+class WikiController(BaseController):
+    """Controls the wiki"""
+
+    def _before(self, *args, **kw):
+        try:
+            repo_path = tg.config.get('wiki.repo')
+            if not repo_path:
+                tg.abort(400, "Wiki not enabled")
+            self.repo = Repository(repo_path)
+        except pg.GitError:
+            self._init_wiki_repo()
+
+    def _init_wiki_repo(self):
+        self.repo = init_repository(tg.config.get('wiki.repo'), True)
+        signature = Signature("Example McPersonson", "emailuser@emailsite.tld") #TODO: Replace me
+
+        tb = self.repo.TreeBuilder()
+
+        # Create blob from frontpage content and insert into tree
+        from pkg_resources import resource_filename
+        filename = "FrontPage.rst"
+        filepath = resource_filename('acmwebsite', 'wiki-assets/FrontPage.rst')
+        blobid = self.repo.create_blob_fromdisk(filepath)
+        tb.insert(filename, blobid, pg.GIT_FILEMODE_BLOB)
+        tree = tb.write()
+
+        # Commit the change
+        branch_commit = self.repo.create_commit(
+            'HEAD',
+            signature,
+            signature,
+            "Initial commit",
+            tree,
+            []
+        )
+
+    @expose('acmwebsite.templates.wiki_view')
+    def _default(self, pagename):
+        """Display a specific page"""
+        settings = {'initial_header_level': 2,
+                    'file_insertion_enabled': 0,
+                    'raw_enabled': 0,
+                    'disable_config': 1,
+        }
+
+        tb = self.repo.TreeBuilder(self.repo.head.peel(Tree))
+        if not tb.get(pagename + '.rst'):
+            tg.abort(404, "Page not found")
+        blob = self.repo.get(self.repo.head.peel(Tree)[pagename + '.rst'].id)
+        document = publish_parts(blob.data, writer_name='html5', settings_overrides=settings)
+        return dict(pagename=pagename, parts=document)
+
+    @expose('acmwebsite.templates.wiki_history')
+    def history(self, pagename):
```

Comment:
Can we do in OO-manner like user profile controller?

Reply:
See f30995d - I attempted to follow the style of the user controller but it still seems a little dirty to me. Let me know what you think.

I moved the repo check/initialization logic from WikiController to WikiPagesController, but I had to change _before to __init__ as it seems WikiPagesController wasn't calling _before like WikiController did.